### PR TITLE
Add a samplerUniforms field for FragmentProgram

### PR DESCRIPTION
**Pull request by chriscraws**
This PR adds support for image samplers to the FragmentProgram shader API.

https://github.com/flutter/flutter/issues/85240

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by flar** [comment]
I'm not familiar enough with spirv to review those files, but the engine parts (lib/ui Dart and C++ classes) look fine.

File: lib/spirv/lib/src/transpiler.dart [hunk 1]
```diff
@@ -147,17 +178,43 @@ class _Transpiler {
...
     // TODO(antrob): Investigate if `List<bool>.filled(maxFunctionId, false)` can be used here instead.
     final Set<int> visited = <int>{};
     writeFunctionAndDeps(visited, entryPoint);
+
+    // Add uniform declarations to header.
+    if (uniformDeclarations.isNotEmpty) {
+      header.writeln();
+      final List<int> locations = uniformDeclarations.keys.toList();
+      locations.sort((int a, int b) => a - b);
+      for (final location in locations) {
```
Comment (zanderso) [inline lib/spirv/lib/src/transpiler.dart:200]: According to the style guide, the `location` declaration should have an explicit type: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-var-and-dynamic.

I thought that would have been caught by an analyzer lint, but we must not have it set up correctly.

File: lib/spirv/lib/src/transpiler.dart [hunk 2]
```diff
@@ -147,17 +178,43 @@ class _Transpiler {
...
+        header.writeln(uniformDeclarations[location]);
+      }
+    }
+
+    // Add SkSL sampler size declarations to header.
+    if (samplerSizeDeclarations.isNotEmpty) {
+      header.writeln();
+      final List<int> locations = samplerSizeDeclarations.keys.toList();
+      locations.sort((int a, int b) => a - b);
+      for (final location in locations) {
```
Comment (zanderso) [inline lib/spirv/lib/src/transpiler.dart:210]: ditto

File: lib/ui/painting.dart [hunk 3]
```diff
@@ -3785,18 +3785,18 @@ class FragmentProgram extends NativeFieldWrapperClass1 {
       spirv,
       spv.TargetLanguage.sksl,
     );
+    if (debugPrint) print(result.src);
```
Comment (zanderso) [inline lib/ui/painting.dart:3788]: The `print` should go on the next line: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#separate-the-if-expression-from-its-statement

**Review by zanderso — APPROVED** [review]
lgtm w/ nits

File: lib/ui/painting.dart (hunk 3 above)
Comment (chriscraws) [reply lib/ui/painting.dart]: Thanks - I meant to delete this actually :)

**Pull request merged**